Wangscape/noise/module/Bitmap.cpp

```diff
@@ -0,0 +1,116 @@
+#include "Bitmap.h"
+#include "logging/Logging.h"
+
+namespace noise
+{
+namespace module
+{
+
+namespace {
+// TODO (Serin): extract the data format from noise::RasterBase into a separate class and use it here
+static std::map<std::string, std::shared_ptr<sf::Image>> loadedImages;
+}
+
+
+Bitmap::Bitmap() : 
+    Module::Module(GetSourceModuleCount())
+{
+}
+
+const std::string & Bitmap::GetFilename() const
+{
+    return mFilename;
+}
+
+void Bitmap::SetFilename(const std::string & filename)
+{
+    // Load and process file if not already done
+    auto it = loadedImages.find(filename);
+    if (it == loadedImages.cend())
+        return;
+    else
+    {
+        auto bitmap = std::make_shared<sf::Image>();
+        // TODO get base file path from Options. How?
+        bitmap->loadFromFile(filename);
+        for (size_t y = 0; y < bitmap->getSize().y; y++)
```

Comment:
I see that's it's the same problem I've faced in my [image-processing](https://github.com/hryniuk/image-processing) project - looping through an image is repeated everywhere and there are a few special operations that you'd like to do on it... I've used instances of `QImage` class there and I've created functions accepting them to do e.g. looping through the whole image with a function (something like `std::for_each`).

In Wangscape, I think, we could have an abstraction over image that will hide the exact type of it and will provide some useful methods as `isGreyscale()`, which could be used in this line. What do you think about it?